### PR TITLE
feat: custom project files

### LINKED ISSUE #656: ⚡️ Feature: Enrich configuration editor
### Feature Description

the project configuration file only accepts `compose.yaml` & `.env` file. when i want to add new file, i had to manually create the file within ssh & adjust the user groups & permissions.

it would be nice if we have custom configuration files other than those two. 

### Problem It Solves

this would help with custom file configuration, other than those files. extending further capabilities in project generation.

### Proposed Solution

i think this could work by adding crud feature for new files.

example: i want to setup `servers.json` to run pgadmin.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you be willing to work on this feature?

- [ ] Yes, I'd like to implement this feature
- [x] I could help with parts of this feature
- [x] No, I'm just suggesting the feature

### LINKED ISSUE #1282: ⚡️ Feature: File editor / File browser
### Feature Description

Add a file browser and editor for at least the project directory.

### Problem It Solves

1. Some docker projects need custom config files. Currently they need to be created via ssh or a file browser.
2. Doing a backup of the persistent data before upgrading the docker image is currently not possible from arcane gui.

### Proposed Solution

Would be super nice to have more options in the tree view of a project. See, create and edit folders and files.
Copy complete folder structures for backups.

Image

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you be willing to work on this feature?

- [ ] Yes, I'd like to implement this feature
- [ ] I could help with parts of this feature
- [x] No, I'm just suggesting the feature

### LINKED ISSUE #1416: ⚡️ Feature: Show .env file editor/files for project sub folders when present.
### Feature Description

When viewing a project with includes and sub folders, show the .env editor when selecting a config in a folder that has one.
In classic mode, show the .env files in the file list

### Problem It Solves

This would allow the user to have full control of their projects

### Proposed Solution

Show the .env editor next to the compose editor if a .env file exists
In classic mode, show the .env files in the file list

### Alternatives Considered

_No response_

### Additional Context

Screenshot of a project with sub-folders:

Image

If possible, show the .env editor,

Image

### Would you be willing to work on this feature?

- [ ] Yes, I'd like to implement this feature
- [ ] I could help with parts of this feature
- [ ] No, I'm just suggesting the feature

### PR DESCRIPTION
**Pull request by cabaucom376**
- Refactored project page with customizable split view that combines the best of both existing modes.
- Add custom files within your projects folder into the editor
- ALLOWED_EXTERNAL_PATHS environment variable for people who understand that they may be introducing security risks
  - ex: ALLOWED_EXTERNAL_PATHS=/app/data/configs/watchtower,/app/data/configs/traefik
  - This allows you to add custom files within these directories. For example /app/data/configs/traefik/traefik.yaml.

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces custom project files functionality, allowing users to add arbitrary files within their project folders and configure external paths via `ALLOWED_EXTERNAL_PATHS`. The implementation includes a comprehensive security layer with path validation, symlink resolution, and boundary checking.

**Major Changes:**
- New centralized security module (`fs_secure.go`) with symlink-aware path validation preventing traversal attacks
- Custom file CRUD operations with database persistence and UI integration
- Split-view editor UI with file tree navigation
- Environment variable `ALLOWED_EXTERNAL_PATHS` for controlled external file access
- Reserved filename protection (compose files, `.env`, `.arcane`) at project root

**Security Implementation:**
- `ValidatePath` function uses `filepath.EvalSymlinks` to resolve symlinks before boundary checks
- Paths validated against both project directory and admin-configured allowed paths
- Comprehensive test coverage including symlink escape and path traversal scenarios
- All file operations go through centralized validation layer

**Architecture:**
The refactoring consolidates previously scattered path validation logic into a single, well-tested security module. Include files and custom files now share the same validation code path, reducing inconsistency risk.

</details>
<details open><summary><h3>Confidence Score: 4/5</h3></summary>

- This PR is generally safe to merge with minor security considerations addressed in previous review threads
- The implementation demonstrates strong security practices with comprehensive path validation, symlink resolution, and extensive test coverage. The centralized security module (`fs_secure.go`) properly handles edge cases like symlink escapes and path traversal. Previous review comments identified legitimate security concerns (symlink validation in one function, error handling) that should be addressed. The feature is well-architected with clear separation of concerns between handlers, services, and security utilities.
- Review previous thread comments on `backend/internal/utils/projects/custom_files.go` for symlink validation and error handling issues
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| backend/internal/utils/fs/fs_secure.go | New centralized path validation module with symlink resolution and boundary checking. Implements defense-in-depth security with `ResolveFilePath`, `IsWithinDirectory`, and `ValidatePath` functions. |
| backend/internal/services/project_service.go | Added three new methods for custom file operations (`CreateProjectCustomFile`, `UpdateProjectCustomFile`, `RemoveProjectCustomFile`). All properly validate paths using the new security module before operations. |
| backend/internal/huma/handlers/projects.go | Added three HTTP endpoints for custom file management. Handlers properly delegate to service layer for validation and operations. |
| frontend/src/routes/(app)/projects/[projectId]/+page.svelte | Implemented split-view editor UI for custom files with state management, file tree navigation, and CRUD operations. Includes validation and error handling for user-provided file paths. |

</details>

</details>

**Context used:**

- Rule from `dashboard` - GoLang Best Practices

---
description: 'Instructions for writing Go code following idiomatic Go pra... ([source](https://app.greptile.com/review/custom-context?memory=214b40a8-9695-4738-986d-5949b5d65ff1))

### CODE REVIEW

**Comment by cabaucom376** [comment]
@greptileai

**Comment by cabaucom376** [comment]
Good feedback from greptile I'll fix it when I get home in an hour

**Comment by cabaucom376** [comment]
@greptileai

**Comment by kmendell** [comment]
I'll try to look at this soon. Im prioritizing bugs over features right now though.

**Comment by cabaucom376** [comment]
Awesome I'll just try and keep it conflict free in the meantime. It's functionally stable now.

**Comment by cabaucom376** [comment]
@greptileai

File: frontend/project.inlang/cache/plugins/2sy648wh9sugi
Comment (kmendell) [inline frontend/project.inlang/cache/plugins/2sy648wh9sugi:1]: Make sure to run pnpm install, to make sure you have the latest paraglide install these files should not be committed paraglide handles a internal git ignore for them

File: backend/internal/config/config.go [hunk 1]
```diff
@@ -53,9 +53,10 @@ type Config struct {
 	GPUMonitoringEnabled    bool   `env:"GPU_MONITORING_ENABLED" default:"false"`
 	GPUType                 string `env:"GPU_TYPE" default:"auto"`
 
-	FilePerm   os.FileMode `env:"FILE_PERM" default:"0644"`
-	DirPerm    os.FileMode `env:"DIR_PERM" default:"0755"`
-	GitWorkDir string      `env:"GIT_WORK_DIR" default:"data/git"`
+	FilePerm             os.FileMode `env:"FILE_PERM" default:"0644"`
+	DirPerm              os.FileMode `env:"DIR_PERM" default:"0755"`
+	AllowedExternalPaths string      `env:"ALLOWED_EXTERNAL_PATHS" default:""`
```
Comment (kmendell) [inline backend/internal/config/config.go]: I would move this to the settings Service and then add the envOverride reflect tag likes others do and then it can be configuredin the ui too.

File: backend/internal/huma/handlers/projects.go [hunk 2]
```diff
@@ -294,6 +324,45 @@ func RegisterProjects(api huma.API, projectService *services.ProjectService) {
...
+			{"ApiKeyAuth": {}},
+		},
+	}, h.UpdateProjectCustomFile)
+
+	huma.Register(api, huma.Operation{
+		OperationID: "remove-project-custom-file",
+		Method:      http.MethodDelete,
+		Path:        "/environments/{id}/projects/{projectId}/custom-files",
+		Summary:     "Remove project custom file",
+		Description: "Remove a custom file from a project (does not delete from disk)",
```
Comment (kmendell) [inline backend/internal/huma/handlers/projects.go]: I may have no gotten there yet but why not remove from disk?

File: backend/internal/services/project_service.go [hunk 3]
```diff
@@ -340,23 +345,25 @@ func (s *ProjectService) GetProjectDetails(ctx context.Context, projectID string
...
-		} else {
-			slog.WarnContext(ctx, "Failed to parse includes", "error", parseErr, "path", projectPath)
-		}
+	composeFile, err := projects.DetectComposeFile(projectPath)
+	if err != nil {
+		return
 	}
+	includes, err := projects.ParseIncludes(composeFile)
+	if err != nil {
+		slog.WarnContext(ctx, "Failed to parse includes", "error", err, "path", projectPath)
```
Comment (kmendell) [inline backend/internal/services/project_service.go]:
```suggestion
		slog.ErrorContext(ctx, "Failed to parse includes", "error", err, "path", projectPath)
```

File: backend/internal/services/project_service.go [hunk 4]
```diff
@@ -340,23 +345,25 @@ func (s *ProjectService) GetProjectDetails(ctx context.Context, projectID string
...
+		return
+	}
+	resp.IncludeFiles = includes
+}
+
+func (s *ProjectService) enrichWithCustomFiles(ctx context.Context, projectPath string, resp *project.Details) {
+	files, err := projects.ParseCustomFiles(projectPath, s.getAllowedExternalPaths(ctx))
+	if err != nil {
+		slog.WarnContext(ctx, "Failed to parse custom files", "error", err, "path", projectPath)
```
Comment (kmendell) [inline backend/internal/services/project_service.go]:
```suggestion
		slog.ErrorContext(ctx, "Failed to parse custom files", "error", err, "path", projectPath)
```

File: backend/internal/services/project_service.go [hunk 5]
```diff
@@ -1153,14 +1160,60 @@ func (s *ProjectService) UpdateProjectIncludeFile(ctx context.Context, projectID
 		return err
 	}
 
-	if err := projects.WriteIncludeFile(proj.Path, relativePath, content); err != nil {
+	if err := projects.WriteIncludeFile(proj.Path, relativePath, content, s.getAllowedExternalPaths(ctx)); err != nil {
```
Comment (kmendell) [inline backend/internal/services/project_service.go]: If the above chnage is done by moving this into the settings service we can then use the settings service to grab the value.

File: backend/internal/utils/projects/files.go [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+	"github.com/goccy/go-yaml"
+)
+
+// ArcaneManifest is the project metadata file, extensible for future features.
+type ArcaneManifest struct {
+	CustomFiles []string `json:"customFiles,omitempty"`
+}
+
+// ArcaneManifestName is the project metadata file name.
+const ArcaneManifestName = ".arcane"
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: Instead of this manifest logic, we should store whatever this manfiest stores in the database as that is faster than reading from disk

File: backend/internal/utils/projects/files.go [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+
+	if checkReserved && withinProject && isReservedFileName(absPath, absProjectDir) {
+		return "", fmt.Errorf("reserved file name: %s", filepath.Base(absPath))
+	}
+
+	return absPath, nil
+}
+
+// validateIncludePath validates an include file path (no reserved name checking).
+func validateIncludePath(projectDir, filePath string, allowedPaths []string) (string, error) {
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: i would gte rid of this and just call the validatePath directly

File: backend/internal/utils/projects/files.go [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+	return validatePath(projectDir, filePath, allowedPaths, false)
+}
+
+// validateCustomFilePath validates a custom file path (with reserved name checking).
+func validateCustomFilePath(projectDir, filePath string, allowedPaths []string) (string, error) {
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: same here

File: backend/internal/utils/projects/files.go [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+
+	return project.IncludeFile{
+		Path:         fullPath,
+		RelativePath: relativePath,
+		Content:      content,
+	}, nil
+}
+
+// WriteIncludeFile writes content to an include file path.
+func WriteIncludeFile(projectDir, includePath, content string, allowedPaths []string) error {
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: We could proabbly make a genric wrapper in utils/fs that would handle all file writing, to reduce duplicate code honestly.

File: backend/internal/utils/projects/files.go [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+	absPath, err := validateIncludePath(projectDir, includePath, allowedPaths)
+	if err != nil {
+		return err
+	}
+	return writeFileWithDir(absPath, content)
+}
+
+// ReadManifest reads the .arcane manifest file.
+func ReadManifest(projectDir string) (*ArcaneManifest, error) {
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: See above about the database

File: backend/internal/utils/projects/files.go [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+	}
+	var m ArcaneManifest
+	if err := json.Unmarshal(data, &m); err != nil {
+		return nil, err
+	}
+	return &m, nil
+}
+
+// WriteManifest writes the .arcane manifest file.
+func WriteManifest(projectDir string, m *ArcaneManifest) error {
```
Comment (kmendell) [inline backend/internal/utils/projects/files.go]: ^^

**Review by kmendell — CHANGES_REQUESTED** [review]
Start with this feedback, as its kind of alot of refactoring, if your confused what i mean just let me know.

File: backend/internal/huma/handlers/projects.go (hunk 2 above)
Comment (cabaucom376) [reply backend/internal/huma/handlers/projects.go]: Ideally you have the choice I guess. I would hate to accidentally delete something that was important and I just didn't want or need to edit in the UI anymore.

Comment (kmendell) [reply backend/internal/huma/handlers/projects.go]: I would follow how the project compose fileds work add a checkbox to remove the files from the disk

**Comment by cabaucom376** [comment]
@kmendell see if that was what you were thinking of and if I interpreted right

File: frontend/src/routes/(app)/projects/[projectId]/+page.svelte [hunk 12]
```diff
@@ -209,17 +318,92 @@
...
 
+	async function handleAddCustomFile() {
+		const trimmedPath = newCustomFileName.trim();
+
+		if (!trimmedPath) {
+			toast.error('Please enter a file name');
+			return;
+		}
+
+		const relativePath = newCustomFileName.trim();
```
Comment (kmendell) [inline frontend/src/routes/(app)/projects/[projectId]/+page.svelte]: We should be using absolute patsh everywhere, for consistency and to avoid the stupid path issuer i screwed up mutiple times

File: backend/internal/huma/handlers/projects.go [hunk 13]
```diff
@@ -294,6 +324,45 @@ func RegisterProjects(api huma.API, projectService *services.ProjectService) {
 		},
 	}, h.UpdateProjectInclude)
 
+	huma.Register(api, huma.Operation{
+		OperationID: "create-project-custom-file",
+		Method:      http.MethodPost,
+		Path:        "/environments/{id}/projects/{projectId}/custom-files",
```
Comment (kmendell) [inline backend/internal/huma/handlers/projects.go]: For sure a nit pick but instead of custom files, i would just chnage it to /files its cleaner

File: frontend/src/lib/components/resizable-split.svelte [hunk 14]
```diff
@@ -1,242 +0,0 @@
...
-	const clampSize = (value: number, minValue: number, maxValue: number) => Math.min(Math.max(value, minValue), maxValue);
-
-	const getAvailableWidth = () => {
-		if (!containerRef) return 0;
-		return Math.max(0, containerRef.getBoundingClientRect().width - handleSize);
-	};
-
-	const getMaxNormalSize = () => Math.max(0, getAvailableWidth() - minSecondSize);
-
-	const ensureInitialSize = () => {
```
Comment (kmendell) [inline frontend/src/lib/components/resizable-split.svelte:66]: Were missing this logic in your new one.

File: frontend/src/routes/(app)/projects/components/CodePanel.svelte
Comment (kmendell) [inline frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Ill be completely honest i dont like how you can reszie all 3. The files panel should be a dynamic width, but should be able to be resized, it also should be colsible i think. like vscode does it .

Comment (kmendell) [inline frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Anotehr thought is to get rid of the panels and use tabs for the open files. but i dont think removing the old classic view is a good idea, i think we leave that and then have the tab bar that shows the files like i had the includes before as not everyone liked this ui.

Comment (kmendell) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Nvm i see you had the context menu, i thinkwhen something oepns though it should split the current viewport, i do think we should keep the legacy ui swith though , maybe in the aperance settings page, as if someone doesnt like this ui they can use the old one.

File: frontend/src/routes/(app)/projects/[projectId]/+page.svelte [hunk 15]
```diff
@@ -746,3 +1021,33 @@
 		</div>
 	</div>
 {/if}
+
+<Dialog.Root bind:open={showAddCustomFileDialog}>
```
Comment (kmendell) [inline frontend/src/routes/(app)/projects/[projectId]/+page.svelte]: I would just add the file into the files panel, no dialog, this would make a better UX experience.

File: frontend/src/routes/(app)/projects/components/CodePanel.svelte
Comment (cabaucom376) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Both the tree view and right panel are collapsible and the right one you can remove it. Its in a way a layout that gives you the best of both worlds. I much prefer the classic UI but the file bar on the top wasn't quit the best especially if you start collecting a lot of files. The tree view panel is much more useful for that. It kind of is just like vscode, you can collapse the left or right one

Comment (cabaucom376) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Can you elaborate some? you can split the current viewport if you want and it is by default.

File: frontend/src/routes/(app)/projects/[projectId]/+page.svelte [hunk 16]
```diff
@@ -214,12 +335,93 @@
...
+		const filePath = newCustomFileName.trim();
+
+		if (!filePath) {
+			cancelAddingCustomFile();
+			return;
+		}
+
+		const result = await tryCatch(projectService.createProjectCustomFile(projectId, filePath));
+		if (result.error) {
+			toast.error(`Failed to add file: ${result.error.message || 'Unknown error'}`);
```
Comment (kmendell) [inline frontend/src/routes/(app)/projects/[projectId]/+page.svelte]: I18n string missing.

File: frontend/src/routes/(app)/projects/[projectId]/+page.svelte [hunk 17]
```diff
@@ -214,12 +335,93 @@
...
+			return;
+		}
+
+		toast.success(`Added ${filePath}`);
```
Comment (kmendell) [inline frontend/src/routes/(app)/projects/[projectId]/+page.svelte]: I18n string missing.

File: frontend/src/routes/(app)/projects/components/CodePanel.svelte
Comment (kmendell) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte]: Do not make the file panele resizable, theres too much going on with that then. Just make it sort of like teh sidebar where its dynamic width based on the item name length, but its not resizable after that, and then it can be collapsed. -- People also did not like this file ui in the past, so for people that just wnat compose and env, we should keep the old classic ui somehow like the switch that i had.

Comment (kmendell) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Theres too much going on with resizing it not intuitive,

Comment (kmendell) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Do not make the file panele resizable, theres too much going on with that then. Just make it sort of like teh sidebar where its dynamic width based on the item name length, but its not resizable after that, and then it can be collapsed. -- People also did not like this file ui in the past, so for people that just wnat compose and env, we should keep the old classic ui somehow like the switch that i had.

Comment (cabaucom376) [reply frontend/src/routes/(app)/projects/components/CodePanel.svelte:1]: Maybe theres just a need for a new approach entirely then? Heres my assessment and opinions about the UX:

- I much prefer having the classic view compose/env side by side and you introducing the resizing between those is for sure a plus because sometimes one can have longer content than the other and it's nice to have that flexibility.
- I don't really like the classic view approach of the files in the top bar as its not really organized and can get unwieldy and hard to quickly navigate when you start introducing more files.
- I would like to maintain the ability to decide wether to have a split view between two files or not, as sometimes with includes there may be a separate env file in a different directory that I would like to edit alongside the separate compose that I am trying to include in the original orchestration one. Also there are sometimes projects that I setup where they're so simple I don't want/need an env file at all, another good reason to allow for the user to not have but one editor panel open.
- I never liked the tree view because of the single editor limitation and therefore found the toggle for the mode in the way and takes up vertical space.
- The tree view is currently resizable and not dynamic width for the filenames as when you get an absolute path like `/srv/containers/config/watchtower/config.json` things just get long and unwieldy so its nice to move the panel smaller or larger to show/hide more while it doesn't get cramped up wrapping multiple lines per file. I do think theres definitely room here though for something better instead of this panel but in all honesty I'm drawing blanks. When I look at vscode and other editors they almost always have a file browser on the left that you can resize and minimize.

So TLDR maybe it seems like the file management UX needs to be reassessed? Heres a few options I can think of:
1. Do as you say and remove the resizability of the file tree.
    - Pros
      - Simple
    - Cons
      - Divisive
      - Mobile? (To be fair I think my mobile solution is trash right now anyway)
2. Remove the large card headers to save vertical space and then chuck the file tree with a vertical resize split over the right panel or something.
    - Pros
      - Flexible
    - Cons
      - Probably too complex, don't even consider this option lol
3. Same thing I have now but you can actually remove the tree view as well (default to no tree view, saving a new include statement triggers it to show). Hover over the left or right margin of the central panel to open a panel (if closed)
    - Pros
      - Lots of flexibility
      - Clean
    - Cons
      - Discoverability? Maybe not too bad but then its out of the way for most people but those who need it can have it.
4. Change the base branch of this PR onto #1535 and then continue to work out a nice file management system for consistency. Add some sort of a ResponsiveDialog triggered by tapping the title of one of the panels for picking/adding a file. Similar split view experience as last option where you hover over the left or right of a panel to get a small drop zone indicator that somewhat pushes the open panel out of the way to click and pick a file to open in split view, max of two open panels instead of 3 since file interactions are a dialog/sheet.
    - Pros
      - Lots of flexibility
      - Very clean
      - Easier to discover with proper indication you can touch the editor title
    - Cons
      - Not really sure need some input
5. Something I'm not thinking of at the moment.

My vote is that I go with option 4. I feel like I can craft something truly clean and minimal while staying powerful for those who need it. I might have to describe my vision better but I would like your input first.

**Comment by neurekadev** [comment]
How is this PR coming along? Custom project files is one of the few things missing for me to fuilly adopt arcane; primarily because some containers want you to map to a file directly and being able to add a file via arcane would mean that I don't have to open an editor or terminal to make changes.

Edit: How would this tie into something like: https://github.com/getarcaneapp/arcane/issues/1282

**Comment by kmendell** [comment]
Closing this for now due to all of the conflicts, Will try to do it for 1.17.0,
